File: orchestrator.py
"""
Orchestrator Agent - 主控制器
- 接收飞书消息
- 意图分类（需求/Bug/查询）
- 编排 Agent 流水线
- 汇总结果并回复飞书
"""

import asyncio
import logging
from typing import Optional, Callable, Awaitable
import toml

from memory_manager import MemoryManager
from state_machine import StateMachine, WorkflowState, TaskContext
from agents import PMAgent, DesignAgent, DevAgent, TestAgent, DeployAgent

logger = logging.getLogger(__name__)


class Orchestrator:
    """
    主编排器。

    外部接口：
    - handle_feishu_message(chat_id, message_id, user_id, text) → 启动流水线
    - confirm_deploy(task_id) → 人工确认部署
    - reject_task(task_id, reason) → 人工拒绝/终止任务
    - list_active_tasks() → 查询进行中的任务
    """

    _BUG_KEYWORDS   = ["bug", "错误", "报错", "问题", "修复", "fix", "崩溃", "异常", "失败"]
    _QUERY_KEYWORDS = ["查询", "进度", "状态", "怎么", "什么", "如何", "帮我查", "查一下"]

    def __init__(
        self,
        config_path: str = "config.toml",
        notify_callback: Optional[Callable[[str, str], Awaitable[None]]] = None,
    ):
        self._cfg = toml.load(config_path)
        self._notify = notify_callback or self._default_notify
        self._memory = MemoryManager(config_path)
        self._state  = StateMachine()
        self._proj_cfg = self._cfg.get("project", {})

        self._pm     = PMAgent(config_path)
        self._design = DesignAgent(config_path)
        self._dev_be = DevAgent(backend=True,  config_path=config_path)
        self._dev_fe = DevAgent(backend=False, config_path=config_path)
        self._test   = TestAgent(config_path)
        self._deploy = DeployAgent(config_path)

    # ------------------------------------------------------------------ #
    # 对外接口
    # ------------------------------------------------------------------ #

    async def handle_feishu_message(
        self,
        chat_id: str,
        message_id: str,
        user_id: str,
        text: str,
    ):
        """接收飞书消息，分类后启动对应流水线"""
        intent = self._classify_intent(text)
        logger.info(f"[Orchestrator] 意图分类: {intent} | 内容: {text[:80]}")

        if intent == "query":
            await self._notify(chat_id, self._answer_query())
            return

        ctx = self._state.create_task(
            original_request=text,
            feishu_chat_id=chat_id,
            feishu_message_id=message_id,
            requester_user_id=user_id,
            max_test_retries=self._proj_cfg.get("max_test_retries", 3),
        )
        ctx.metadata["intent"] = intent

        label = "需求" if intent == "feature" else "Bug修复"
        await self._notify(
            chat_id,
            f"✅ 已收到{label}请求，任务 ID：`{ctx.task_id}`\n正在启动 AI 流水线，请稍候...",
        )
        asyncio.create_task(self._run_pipeline(ctx))

    async def confirm_deploy(self, task_id: str, user_id: str):
        """人工确认部署（飞书按钮触发）"""
        ctx = self._state.load_task(task_id)
        if not ctx:
            return
        if ctx.state != WorkflowState.DEPLOY_CONFIRM.value:
            await self._notify(ctx.feishu_chat_id, f"❌ 任务 {task_id} 当前状态不是等待部署确认")
            return
        ctx = self._state.transition(ctx, WorkflowState.DEPLOYING)
        ctx.metadata["confirmed_by"] = user_id
        asyncio.create_task(self._run_deploy(ctx))

    async def reject_task(self, task_id: str, reason: str = ""):
        """人工拒绝/终止任务"""
        ctx = self._state.load_task(task_id)
        if ctx:
            self._state.fail_task(ctx, f"人工终止: {reason}")
            await self._notify(ctx.feishu_chat_id, f"⛔ 任务 {task_id} 已终止。{reason}")

    def list_active_tasks(self) -> list[TaskContext]:
        """查询所有进行中的任务（供 FeishuBot 等外部调用）"""
        return self._state.list_active_tasks()

    # ------------------------------------------------------------------ #
    # 内部：流水线
    # ------------------------------------------------------------------ #

    async def _run_pipeline(self, ctx: TaskContext):
        """完整的 PLANNING → DESIGN → DEVELOPMENT → TESTING 流水线"""
        try:
            # 整个流水线共享同一次 recall，避免重复查询
            memory_ctx = self._memory.recall(ctx.original_request)

            # ── PLANNING ──────────────────────────────────────────────
            ctx = self._state.transition(ctx, WorkflowState.PLANNING)
            await self._notify(ctx.feishu_chat_id, f"📋 [{ctx.task_id}] PM Agent 正在分析需求...")
            story_path = await asyncio.to_thread(
                self._pm.analyze, ctx.original_request, memory_ctx, ctx.task_id,
            )
            ctx = self._state.transition(ctx, WorkflowState.DESIGN, story_path=story_path)

            # ── DESIGN ────────────────────────────────────────────────
            await self._notify(ctx.feishu_chat_id, f"🏗️ [{ctx.task_id}] Design Agent 正在设计技术方案...")
            tech_spec_path = await asyncio.to_thread(
                self._design.design, story_path, memory_ctx, ctx.task_id,
            )
            ctx = self._state.transition(ctx, WorkflowState.DEVELOPMENT, tech_spec_path=tech_spec_path)
            await self._notify(
                ctx.feishu_chat_id,
                f"📐 [{ctx.task_id}] 技术方案已生成：`{tech_spec_path}`\n🚀 开始代码实现...",
            )

            # ── DEVELOPMENT ───────────────────────────────────────────
            be_report, fe_report = await self._run_development(ctx, tech_spec_path, memory_ctx)
            ctx = self._state.transition(
                ctx, WorkflowState.TESTING, dev_report_path=be_report,
            )
            await self._notify(ctx.feishu_chat_id, f"🧪 [{ctx.task_id}] 开始运行测试...")

            # ── TESTING（带重试）─────────────────────────────────────
            await self._run_testing_loop(ctx, memory_ctx)

        except Exception as e:
            logger.exception(f"[Orchestrator] 流水线异常: {e}")
            self._state.fail_task(ctx, str(e))
            await self._notify(
                ctx.feishu_chat_id,
                f"❌ [{ctx.task_id}] 流水线异常：{e}\n请查看日志排查问题。",
            )

    async def _run_development(
        self, ctx: TaskContext, tech_spec_path: str, memory_ctx: str
    ) -> tuple[str, str]:
        """并行运行前后端 Dev Agent"""
        be_task = asyncio.to_thread(
            self._dev_be.implement, tech_spec_path, memory_ctx, ctx.task_id,
        )
        fe_task = asyncio.to_thread(
            self._dev_fe.implement, tech_spec_path, memory_ctx, ctx.task_id,
        )
        be_report, fe_report = await asyncio.gather(be_task, fe_task)
        await self._notify(
            ctx.feishu_chat_id,
            f"✍️ [{ctx.task_id}] 代码实现完成\n- 后端报告: `{be_report}`\n- 前端报告: `{fe_report}`",
        )
        return be_report, fe_report

    async def _run_testing_loop(self, ctx: TaskContext, memory_ctx: str):
        """测试循环：最多重试 max_test_retries 次"""
        test_command = self._proj_cfg.get("test_command", "make test")

        while True:
            ctx = self._state.load_task(ctx.task_id)
            retry = ctx.test_retry_count

            passed, report_path = await asyncio.to_thread(
                self._test.run_tests,
                ctx.story_path,
                ctx.dev_report_path,
                test_command,
                memory_ctx,
                ctx.task_id,
                retry,
            )
            ctx = self._state.transition(ctx, WorkflowState.TESTING, test_report_path=report_path)

            if passed:
                await self._notify(
                    ctx.feishu_chat_id,
                    f"✅ [{ctx.task_id}] 测试全部通过！\n\n"
                    f"是否部署到生产？请回复：\n"
                    f"• **确认部署** `/deploy {ctx.task_id}`\n"
                    f"• **取消** `/cancel {ctx.task_id}`",
                )
                self._state.transition(ctx, WorkflowState.DEPLOY_CONFIRM)
                self._memory.write_daily_log(
                    "orchestrator",
                    f"任务 {ctx.task_id} 测试通过，等待部署确认。需求：{ctx.original_request[:100]}",
                )
                return

            self._state.increment_test_retry(ctx)

            if not self._state.can_retry_test(ctx):
                ctx = self._state.transition(ctx, WorkflowState.REVIEW)
                await self._notify(
                    ctx.feishu_chat_id,
                    f"🚨 [{ctx.task_id}] 测试连续失败 {ctx.max_test_retries} 次，需要人工介入！\n"
                    f"测试报告：`{report_path}`\n"
                    f"• `/reject {ctx.task_id} <原因>` - 终止任务",
                )
                return

            feedback_path = self._test.get_feedback_path(ctx.task_id)
            await self._notify(
                ctx.feishu_chat_id,
                f"⚠️ [{ctx.task_id}] 第 {retry + 1} 轮测试失败，Dev Agent 正在修复...",
            )
            ctx = self._state.transition(ctx, WorkflowState.DEVELOPMENT)
            await asyncio.to_thread(
                self._dev_be.run,
                f"测试失败，请根据以下反馈修复代码：\n\n反馈文件：`{feedback_path}`\n\n"
                f"读取反馈文件，找到并修复对应的代码问题。修复完成后更新开发报告。",
            )
            ctx = self._state.transition(ctx, WorkflowState.TESTING)

    async def _run_deploy(self, ctx: TaskContext):
        """执行部署"""
        deploy_command = self._proj_cfg.get("deploy_command", "git push origin main")
        try:
            success, report_path = await asyncio.to_thread(
                self._deploy.deploy,
                ctx.test_report_path,
                deploy_command,
                ctx.task_id,
                confirmed=True,
            )
            if success:
                ctx = self._state.transition(ctx, WorkflowState.DONE, deploy_report_path=report_path)
                await self._notify(
                    ctx.feishu_chat_id,
                    f"🎉 [{ctx.task_id}] 部署成功！\n需求「{ctx.original_request[:60]}...」已上线。",
                )
                self._memory.write_daily_log(
                    "orchestrator",
                    f"任务 {ctx.task_id} 部署成功。需求：{ctx.original_request[:100]}",
                )
            else:
                self._state.fail_task(ctx, "部署失败")
                await self._notify(
                    ctx.feishu_chat_id,
                    f"❌ [{ctx.task_id}] 部署失败！\n报告：`{report_path}`\n请人工检查 CI/CD 状态。",
                )
        except Exception as e:
            self._state.fail_task(ctx, str(e))
            await self._notify(ctx.feishu_chat_id, f"❌ [{ctx.task_id}] 部署异常：{e}")

    # ------------------------------------------------------------------ #
    # 意图分类
    # ------------------------------------------------------------------ #

    def _classify_intent(self, text: str) -> str:
        lower = text.lower()
        if any(kw in lower for kw in self._QUERY_KEYWORDS):
            return "query"
        if any(kw in lower for kw in self._BUG_KEYWORDS):
            return "bug"
        return "feature"

    def _answer_query(self) -> str:
        active_tasks = self._state.list_active_tasks()
        if not active_tasks:
            return "当前没有进行中的任务。"
        lines = ["当前进行中的任务："]
        for t in active_tasks:
            lines.append(f"• `{t.task_id}` [{t.state}] {t.original_request[:50]}...")
        return "\n".join(lines)

    async def _default_notify(self, chat_id: str, text: str):
        logger.info(f"[飞书通知] chat={chat_id}: {text}")
        print(f"\n[飞书] {text}\n")

File: src/orchestrator.ts
/**
 * Orchestrator - 主控制器
 * 编排 PM → Design → Dev → Test → Deploy 流水线
 */

import { AgentMemoryClient } from './memory_client.js';
import { StateMachine, WorkflowState, TaskContext } from './state_machine.js';
import { AppConfig } from './config.js';
import { analyzePM } from './agents/pm_agent.js';
import { runDesign } from './agents/design_agent.js';
import { runDev, runDevFix } from './agents/dev_agent.js';
import { runTests, getFeedbackPath } from './agents/test_agent.js';
import { runDeploy } from './agents/deploy_agent.js';

type IntentType = 'feature' | 'bug' | 'query' | 'invalid';

const VALID_INTENTS = new Set<IntentType>(['feature', 'bug', 'query', 'invalid']);

const INTENT_SYSTEM_PROMPT = `你是意图分类器。将用户消息分类为以下类别之一：
- feature：新功能需求或开发任务
- bug：Bug修复、错误处理请求
- query：查询任务状态、进度、结果
- invalid：问候、测试、无意义内容、太短的消息`;

export type NotifyPayload =
  | { kind: 'text'; text: string }
  | { kind: 'deploy_card'; taskId: string; summary: string };

export type NotifyCallback = (chatId: string, payload: NotifyPayload) => Promise<void>;

export class Orchestrator {
  private cfg: AppConfig;
  private notify: NotifyCallback;
  private memory: AgentMemoryClient;
  private state: StateMachine;
  private cwd: string;
  private blackboxApiKey: string;

  constructor(cfg: AppConfig, notifyCallback?: NotifyCallback) {
    this.cfg = cfg;
    this.notify = notifyCallback ?? this.defaultNotify.bind(this);
    this.memory = new AgentMemoryClient();
    this.state = new StateMachine();
    this.cwd = cfg.project.workspace_path;
    this.blackboxApiKey = cfg.blackbox.api_key;
    if (!this.blackboxApiKey) console.warn('[Orchestrator] BLACKBOX_API_KEY 未配置，意图分类将失败');
  }

  // ── 便捷方法 ──────────────────────────────────────────────────────── //

  private text(chatId: string, msg: string): Promise<void> {
    return this.notify(chatId, { kind: 'text', text: msg });
  }

  // ── 对外接口 ──────────────────────────────────────────────────────── //

  async handleFeishuMessage(
    chatId: string,
    messageId: string,
    userId: string,
    text: string,
  ): Promise<void> {
    let intent: IntentType;
    try {
      intent = await this.classifyIntent(text);
    } catch (e) {
      console.error(`[Orchestrator] 意图分类失败: ${e}`);
      await this.text(chatId, `⚠️ 意图识别暂时不可用，请稍后重试。`);
      return;
    }
    console.log(`[Orchestrator] 意图分类: ${intent} | 内容: ${text.slice(0, 80)}`);

    if (intent === 'invalid') {
      await this.text(chatId, '👋 你好！请描述你的需求或 Bug，我会为你启动 AI 开发流水线。\n\n例如：\n• "实现用户登录功能"\n• "修复首页加载慢的问题"');
      return;
    }

    if (intent === 'query') {
      await this.text(chatId, this.answerQuery());
      return;
    }

    const ctx = this.state.createTask({
      original_request: text,
      feishu_chat_id: chatId,
      feishu_message_id: messageId,
      requester_user_id: userId,
      max_test_retries: this.cfg.project.max_test_retries,
    });
    ctx.metadata['intent'] = intent;

    const label = intent === 'feature' ? '需求' : 'Bug修复';
    await this.text(
      chatId,
      `✅ 已收到${label}请求，任务 ID：\`${ctx.task_id}\`\n正在启动 AI 流水线，请稍候...`,
    );

    // 异步启动，不 await，避免阻塞飞书响应
    this.runPipeline(ctx).catch(e =>
      console.error(`[Orchestrator] 流水线异常: ${e}`),
    );
  }

  async confirmDeploy(taskId: string, userId: string): Promise<void> {
    const ctx = this.state.loadTask(taskId);
    if (!ctx) return;
    if (ctx.state !== WorkflowState.DEPLOY_CONFIRM) {
      await this.text(ctx.feishu_chat_id, `❌ 任务 ${taskId} 当前状态不是等待部署确认`);
      return;
    }
    const updated = this.state.transition(ctx, WorkflowState.DEPLOYING);
    updated.metadata['confirmed_by'] = userId;
    this.runDeploy(updated).catch(e =>
      console.error(`[Orchestrator] 部署异常: ${e}`),
    );
  }

  async rejectTask(taskId: string, reason = ''): Promise<void> {
    const ctx = this.state.loadTask(taskId);
    if (ctx) {
      this.state.failTask(ctx, `人工终止: ${reason}`);
      await this.text(ctx.feishu_chat_id, `⛔ 任务 ${taskId} 已终止。${reason}`);
    }
  }

  listActiveTasks(): TaskContext[] {
    return this.state.listActiveTasks();
  }

  // ── 内部：流水线 ─────────────────────────────────────────────────── //

  private async runPipeline(ctx: TaskContext): Promise<void> {
    try {
      const memoryCtx = await this.memory.buildContext(ctx.original_request);

      // ── PLANNING ──────────────────────────────────────────────────
      ctx = this.state.transition(ctx, WorkflowState.PLANNING);
      await this.text(ctx.feishu_chat_id, `📋 [${ctx.task_id}] PM Agent 正在分析需求...`);
      void this.memory.setWorking('pm_agent', { taskId: ctx.task_id, status: 'running' });
      const storyPath = await analyzePM(ctx.original_request, memoryCtx, ctx.task_id, this.cwd, this.memory);
      ctx = this.state.transition(ctx, WorkflowState.DESIGN, { story_path: storyPath });

      // ── DESIGN ────────────────────────────────────────────────────
      await this.text(ctx.feishu_chat_id, `🏗️ [${ctx.task_id}] Design Agent 正在设计技术方案...`);
      void this.memory.setWorking('design_agent', { taskId: ctx.task_id, status: 'running', storyPath });
      const techSpecPath = await runDesign(storyPath, memoryCtx, ctx.task_id, this.cwd, this.memory);
      ctx = this.state.transition(ctx, WorkflowState.DEVELOPMENT, { tech_spec_path: techSpecPath });
      await this.text(
        ctx.feishu_chat_id,
        `📐 [${ctx.task_id}] 技术方案已生成：\`${techSpecPath}\`\n🚀 开始代码实现...`,
      );

      // ── DEVELOPMENT ───────────────────────────────────────────────
      void this.memory.setWorking('dev_agent', { taskId: ctx.task_id, status: 'running', techSpecPath });
      const [beReport, feReport] = await Promise.all([
        runDev(techSpecPath, memoryCtx, ctx.task_id, this.cwd, true, this.memory),
        runDev(techSpecPath, memoryCtx, ctx.task_id, this.cwd, false, this.memory),
      ]);
      ctx = this.state.transition(ctx, WorkflowState.TESTING, { dev_report_path: beReport, frontend_report_path: feReport });
      await this.text(
        ctx.feishu_chat_id,
        `✍️ [${ctx.task_id}] 代码实现完成\n- 后端报告: \`${beReport}\`\n- 前端报告: \`${feReport}\``,
      );

      // ── TESTING（带重试）─────────────────────────────────────────
      await this.runTestingLoop(ctx, memoryCtx);
    } catch (e) {
      console.error(`[Orchestrator] 流水线异常: ${e}`);
      // 重新加载最新 ctx，避免使用流水线中途被覆盖的过期引用
      const latestCtx = this.state.loadTask(ctx.task_id) ?? ctx;
      this.state.failTask(latestCtx, String(e));
      await this.text(
        latestCtx.feishu_chat_id,
        `❌ [${latestCtx.task_id}] 流水线异常：${e}\n请查看日志排查问题。`,
      );
    }
  }

  private async runTestingLoop(ctx: TaskContext, memoryCtx: string): Promise<void> {
    const testCommand = this.cfg.project.test_command;

    while (true) {
      const retry = ctx.test_retry_count;

      const { passed, reportPath } = await runTests({
        storyPath: ctx.story_path ?? '',
        devReportPath: ctx.dev_report_path ?? '',
        testCommand,
        memoryContext: memoryCtx,
        taskId: ctx.task_id,
        cwd: this.cwd,
        retryCount: retry,
        memory: this.memory,
      });
      ctx = this.state.transition(ctx, WorkflowState.TESTING, { test_report_path: reportPath });

      if (passed) {
        const summary = `需求「${ctx.original_request.slice(0, 80)}」已通过全部测试，等待确认部署到生产。`;
        await this.notify(ctx.feishu_chat_id, {
          kind: 'deploy_card',
          taskId: ctx.task_id,
          summary,
        });
        ctx = this.state.transition(ctx, WorkflowState.DEPLOY_CONFIRM);
        void this.memory.reflect(ctx.task_id);
        void this.memory.store(
          `任务 ${ctx.task_id} 测试通过，等待部署确认。需求：${ctx.original_request.slice(0, 100)}`,
          'episodic',
          { taskId: ctx.task_id },
        );
        return;
      }

      ctx = this.state.incrementTestRetry(ctx);

      if (!this.state.canRetryTest(ctx)) {
        ctx = this.state.transition(ctx, WorkflowState.REVIEW);
        await this.text(
          ctx.feishu_chat_id,
          `🚨 [${ctx.task_id}] 测试连续失败 ${ctx.max_test_retries} 次，需要人工介入！\n测试报告：\`${reportPath}\`\n• \`/cancel ${ctx.task_id} <原因>\` - 终止任务`,
        );
        return;
      }

      const feedbackPath = getFeedbackPath(ctx.task_id, this.cwd);
      await this.text(
        ctx.feishu_chat_id,
        `⚠️ [${ctx.task_id}] 第 ${retry + 1} 轮测试失败，Dev Agent 正在修复...`,
      );
      ctx = this.state.transition(ctx, WorkflowState.DEVELOPMENT);
      await Promise.all([
        runDevFix(feedbackPath, ctx.task_id, this.cwd, true, this.memory),
        runDevFix(feedbackPath, ctx.task_id, this.cwd, false, this.memory),
      ]);
      ctx = this.state.transition(ctx, WorkflowState.TESTING);
    }
  }

  private async runDeploy(ctx: TaskContext): Promise<void> {
    const deployCommand = this.cfg.project.deploy_command;
    try {
      const { success, reportPath } = await runDeploy({
        testReportPath: ctx.test_report_path ?? '',
        deployCommand,
        taskId: ctx.task_id,
        cwd: this.cwd,
        confirmed: true,
        memory: this.memory,
      });
      if (success) {
        this.state.transition(ctx, WorkflowState.DONE, { deploy_report_path: reportPath });
        await this.text(
          ctx.feishu_chat_id,
          `🎉 [${ctx.task_id}] 部署成功！\n需求「${ctx.original_request.slice(0, 60)}...」已上线。`,
        );
        void this.memory.store(
          `任务 ${ctx.task_id} 部署成功。需求：${ctx.original_request.slice(0, 100)}`,
          'ci_cd',
          { taskId: ctx.task_id },
        );
      } else {
        this.state.failTask(ctx, '部署失败');
        await this.text(
          ctx.feishu_chat_id,
          `❌ [${ctx.task_id}] 部署失败！\n报告：\`${reportPath}\`\n请人工检查 CI/CD 状态。`,
        );
      }
    } catch (e) {
      this.state.failTask(ctx, String(e));
      await this.text(ctx.feishu_chat_id, `❌ [${ctx.task_id}] 部署异常：${e}`);
    }
  }

  // ── 意图分类 ─────────────────────────────────────────────────────── //

  private async classifyIntent(text: string): Promise<IntentType> {
    if (this.blackboxApiKey) {
      try {
        return await this.classifyByBlackbox(text);
      } catch (e) {
        console.warn(`[Orchestrator] Blackbox API 失败，降级为关键词匹配: ${e}`);
      }
    } else {
      console.warn('[Orchestrator] BLACKBOX_API_KEY 未配置，使用关键词匹配兜底');
    }
    return this.classifyByKeywords(text);
  }

  private async classifyByBlackbox(text: string): Promise<IntentType> {
    const { endpoint, model, timeout_ms } = this.cfg.blackbox;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout_ms);

    let res: Response;
    try {
      res = await fetch(endpoint, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.blackboxApiKey}`,
          'Content-Type': 'application/json',
        },
        signal: controller.signal,
        body: JSON.stringify({
          model,
          messages: [
            { role: 'system', content: INTENT_SYSTEM_PROMPT },
            { role: 'user', content: text },
          ],
          tools: [{
            type: 'function',
            function: {
              name: 'set_intent',
              description: '设置意图分类结果',
              parameters: {
                type: 'object',
                properties: {
                  intent: { type: 'string', enum: ['feature', 'bug', 'query', 'invalid'] },
                },
                required: ['intent'],
              },
            },
          }],
          tool_choice: { type: 'function', function: { name: 'set_intent' } },
        }),
      });
    } finally {
      clearTimeout(timer);
    }

    if (!res.ok) {
      let detail = `HTTP ${res.status}`;
      try {
        const err = await res.json() as { error?: { message?: string } };
        detail = err.error?.message ?? detail;
      } catch { /* ignore */ }
      throw new Error(`Blackbox API 错误: ${detail}`);
    }

    const data = await res.json() as {
      choices?: Array<{ message?: { tool_calls?: Array<{ function?: { arguments?: string } }> } }>;
    };
    const raw = data.choices?.[0]?.message?.tool_calls?.[0]?.function?.arguments;
    if (!raw) throw new Error('Blackbox API 返回结构异常：缺少 tool_calls.arguments');

    let args: { intent?: string };
    try {
      args = JSON.parse(raw) as { intent?: string };
    } catch {
      throw new Error(`Blackbox API 返回的 arguments 不是合法 JSON: ${raw}`);
    }

    if (!VALID_INTENTS.has(args.intent as IntentType)) {
      throw new Error(`Blackbox API 返回了非法意图值: ${args.intent}`);
    }
    return args.intent as IntentType;
  }

  /** 关键词兜底分类，无需外部依赖 */
  private classifyByKeywords(text: string): IntentType {
    const t = text.toLowerCase();
    if (t.length < 4) return 'invalid';
    if (/修复|bug|报错|错误|崩溃|异常|失败|不对|问题/.test(t)) return 'bug';
    if (/状态|进度|结果|查询|怎么样|完成了|多少|有没有/.test(t)) return 'query';
    if (/实现|开发|新增|添加|功能|需求|做|帮我|支持|接入/.test(t)) return 'feature';
    return 'invalid';
  }

  private answerQuery(): string {
    const activeTasks = this.state.listActiveTasks();
    if (activeTasks.length === 0) return '当前没有进行中的任务。';
    const lines = ['当前进行中的任务：'];
    for (const t of activeTasks) {
      lines.push(`• \`${t.task_id}\` [${t.state}] ${t.original_request.slice(0, 50)}...`);
    }
    return lines.join('\n');
  }

  private async defaultNotify(chatId: string, payload: NotifyPayload): Promise<void> {
    if (payload.kind === 'text') {
      console.log(`\n[飞书] chat=${chatId}: ${payload.text}\n`);
    } else {
      console.log(`\n[飞书] chat=${chatId}: [部署确认卡片] task=${payload.taskId} ${payload.summary}\n`);
    }
  }
}
